[iOS][expo-av] Fixed black screen bug that appears when playing videos with AV1 codec

HezekielT opened the pull request:
# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

Currently, Apple doesn't fully support videos with AV1 codecs which means when playing those videos on iOS Native devices, the screen becomes black while the audio track plays and no error is thrown which makes it difficult to handle such videos for developers.

More details about AV1 Playback support can be found [here](https://bitmovin.com/blog/av1-playback-support/#av-1-playback-support-news-in-2024)

> Generally speaking, the Chrome browser and Android ecosystem handle AV1 well across phones, tablets, smart TVs and set-top boxes/streaming sticks. Unfortunately, the same cannot be said for Safari and iOS where support had been lacking until the iPhone 15 Pro announcement.

# How

<!--
How did you build this feature or fix this bug and why?
-->

[AVAsset](https://developer.apple.com/documentation/avfoundation/avasset) class has a property called `isPlayable` which can be used to determine if a video asset can be played or not by the device. 

This PR tries to fix the black screen bug by calling `isPlayable` on the video asset inside [`_loadNewPlayer(EXAVPlayerData.m)`](https://github.com/expo/expo/blob/main/packages/expo-av/ios/EXAV/EXAVPlayerData.m#L157-L180) and throwing an error if it is false(!asset.isPlayable) which can then be handled by the developer. 

# Test Plan

<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

1. Add a remote or local video asset inside [`VideoScreen.tsx`](https://github.com/expo/expo/blob/6bfa60a61ef5720ce73d8ce3ce7c061f94602866/apps/native-component-list/src/screens/Audio/AV/VideoScreen.tsx#L11-L23)

Here is a sample video with AV1 codec for testing purpose - https://github.com/user-attachments/assets/2b502b64-83c3-4b81-bc7b-b57cf6892ab0

2. run `yarn ios`
3. When the bare-expo app launches, click on `Component` and scroll down and click on `Video (expo-av)`.
4. Play the AV1 video.
5. Verify the screen remains black while the audio track is playing.

https://github.com/user-attachments/assets/bb30a9cc-8ba3-461b-97d3-4d05081df23e



# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

HezekielT committed:
added a logic that checks if an asset is playable using isPlayable in EXAVPlayerData.m(_loadNewPlayer)

HezekielT committed:
added changelog entry

HezekielT committed:
Merge branch 'expo:main' into fix--iOS-black-screen-when-playing-videos-with-AV1-codec

behenate commented on packages/expo-av/ios/EXAV/EXAVPlayerData.m:
```diff
@@ -158,23 +158,31 @@ - (void)_loadNewPlayer
+      NSString *errorMessage = @"Load encountered an error: [AVAsset isPlayable:] returned false. The asset does not contains a playable content or is not supported by the device.";
       [self _finishLoadWithError:errorMessage];
+      return;
+    } else {
```
We can remove the `else` block, since we have the `return` inside of the `if` block.


```suggestion
    }
```

HezekielT replied:
Thanks for the review @behenate . I have pushed the changes you suggested.

lukmccall committed:
[notifications][ios] Fixed incorrect `EXNotifications-Swift.h` import (#34987)

behenate reviewed (changes requested):
Thanks for the PR! Just a note, we are replacing `expo-av` with `expo-video` and don't have plans for further support of `expo-av`. We will still release this fix in the next SDK version. I left one small suggestion.

I think we may have the same issue in `expo-video` so I'll have to look into it 😄

jonsamp committed:
[docs] Adds Vexo to monitoring services (#34981)

behenate committed:
[video][android][ios] Add support for listing available video tracks (#33458)

# Why

There were requests for the `naturalSize` prop from `expo-av`, which is
currently missing in `expo-video`. In `expo-av` it returned the
resolution being currently played back in the viewport. I have decided
that this is not enough, as each video can be a HLS source having
multiple available resolutions, frameRates, bitrates etc.

# How

- I created a `VideoTrack` record with all necessary information about a
video track. For Android it will be very useful for selecting a specific
track in the future, on iOS it won't be possible but the `bitrate`
property will be useful, as it is possible to partially control the
track being played back by capping the available resolution.
- Introduced `sourceLoad` (the naming is debatable, I'm very open to
suggestions), which contains all information that should be available
once a video track is loaded. This will also make the
`availableSubtitleTracksChange` event redundant, so we can mark is as
deprecated for SDK53 and remove it in a future SDK
- On Android it was relatively easy to create the VideoTracks
- On iOS for regular sources it's also easy, but for HLS it was a bit of
a hell to do it. For HLS we:
- We fetch the playlist file and parse it to obtain information about
all available video tracks.
- When a video track changes (AVKit sends an event but doesn't provide
any data) we can parse the last `AVPlayerItem` log entry to obtain a uri
to which it connected.
- The uri contains an id, which we previously parsed while saving the
playlist file, therefore we can associate it with a VideoTrack.

We now have SubtitleTracks and VideoTracks. In the future we will also
add AudioTracks, to finish the "trio" and allow for switching the
current audio language.

# Test Plan

Tested in BareExpo on Android and iOS on HLS and non-HLS sources


https://github.com/user-attachments/assets/6f918f27-4f65-49b0-bc71-3efcb0868c32

gabrieldonadel committed:
Bump minimum macOS version to 11.0 (#34980)

# Why

The new React Native macOS 0.77 requires the macOS version to be at
least 11.0

# How

- Updates minimum macOS versions in our podspecs
- Update macOS deployment target in BareExpo  
- Clean up code availability for macOS 11.0

# Test Plan

Run BareExpo macOS locally (with some additional patches)

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [ ] I added a `changelog.md` entry and rebuilt the package sources
according to [this short
guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

Kudo committed:
[packages] fix @react-navigation/core resolution warning (#33829)

# Why

fixes outdated resolutions not updated in #33758
```
[1/4] 🔍  Resolving packages...
warning Resolution field "@react-navigation/core@7.0.0" is incompatible with requested version "@react-navigation/core@^7.3.1"
warning Resolution field "@react-navigation/routers@7.0.0" is incompatible with requested version "@react-navigation/routers@^7.1.2"
```

# How

- [monorepo] update core and routers to latest
- [dev-launcher] remove unused core depencencies

Kudo committed:
[fingerprint] prevent null cli options overwrite config options (#34935)

# Why

fixes #34930

# How

filter out null options passing from cli to prevent overwrite config options

HezekielT committed:
remove unnecessary else statement

jakex7 committed:
[Localization] Add languageScriptCode (#34984)

# Why

To be able to differentiate between various writing systems. Resolves
https://github.com/expo/expo/issues/34979

# How

Adds `languageScriptCode`, an [ISO 15924](https://en.wikipedia.org/wiki/ISO_15924) 
4-letter script code. On Android and Web, it may be `null` if none is defined.

![image](https://github.com/user-attachments/assets/03545425-c7ff-432f-9ded-07e28a6ec8eb)

# Test Plan

Bare Expo -> `LocalizationScreen`

# Checklist

- [x] I added a `changelog.md` entry and rebuilt the package sources
according to [this short
guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).
- [x] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

reichhartd committed:
chore: Standardize platform key ordering in `expo-module.config.json` (#35003)

chrfalch committed:
[ios][location] add error handler to streaming location/device methods (#35004)

# Why

When using the streaming location/heading methods (`watchHeadingAsync` /
`watchPositionAsync`) from JS, we're not handling errors very well.

It is possible for the location manager to stop being able to provide
location while we're streaming. This can be tested by changing the
`Features` -> `Location` to `None` while streaming location using f.ex.
`City ride` in the simulator. This is a use case that can happen on a
real device as well.

When this happens, an exception is thrown on the native side. This
exception is not handled by the JS code in any way since all the
location/heading info is delivered through event emitters.

# How

This commit fixes the above by adding an optional errorHandler callback
for the above methods, and by implementing a catch clause on the native
side that will emit an error that will be caught in the JS code and
passed back to the caller.

In addition there was an error with the streamer (location/heading) in
swift causing them to not stop emitting after we close the subscription.
The solution here was to do as on Android, to explicitly stop the
streamers when we unsubscribe (on the native side).

I also refactored the Location methods to better handle errors, both
when initially starting streaming, but also when requesting heading
through `getHeadingAsync`.

Since the new error-handler methods are optional parameters to the
respective functions, it shouldn't be necessary to mark this as a
breaking change.

closes #34714

# Test Plan

Tested in BareExpo

# Checklist

- [x] I added a `changelog.md` entry and rebuilt the package sources
according to [this short
guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).
- [x] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

---------

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>
Co-authored-by: Łukasz Kosmaty <lukasz.kosmaty@swmansion.com>

lukmccall committed:
[bare][ios] Make it work with `use_frameworks!` (#35005)

lukmccall committed:
[expo][ios] Move `AppDelegate` integration from `e-m-c` to `expo` (#34985)

Kudo committed:
[dev-launcher][dev-menu] remove remote debugging dead code (#34977)

# Why

close ENG-14394
fixes #33371

# How

- [dev-launcher][dev-menu] remove remote debugging dead code
- [dev-launcher] since this feature is deprecated for a while and not necessarily to support for core. rather than copy source sets of `NonFinalBridgeDevSupportManager` from rn77 to rn79. i just apply the [patch](https://github.com/facebook/react-native/commit/9aae84a688b5af87faf4b68676b6357de26f797f) to the original `NonFinalBridgeDevSupportManager` for rn77.

Kudo committed:
[go] remove remote debugging dead code (#34978)

# Why

close ENG-14394
fixes #33371

# How

remove remote debugging dead code in expo go

Kudo committed:
[go] fix downloadAsset build error (#34994)

# Why

fixes ios expo-go build error: https://github.com/expo/expo/actions/runs/13363533171

# How

`downloadAsset` interface was changed since #34453. this pr tries to apply the same change from RemoteAppLoader to HomeAppLoader

Kudo committed:
[create-expo-nightly] fix expo/expo yarn install error (#34995)

# Why

fixes create-expo-nightly errors on ci

# How

only show warning for the `yarn install` failure

matejkriz committed:
[fingerprint] fix title in README (#35006)

alanjhughes committed:
[expo-ui] Add DateTimePicker component (#34883)

lukmccall committed:
[native-tests] Run `pod install`

gabrieldonadel committed:
[modules-core][iOS] Fix initializing view definitions on Paper (#35007)

# Why

Follow up of https://github.com/expo/expo/pull/34293

When using the default initializer to create module instances with Paper
(old arch) the viewDefinition does not get defined

# How

Update `ViewModuleWrapper` initializer to use the `DEFAULT_MODULE_VIEW`
view from the module as the default view

# Test Plan

Run BareExpo on old arch

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [ ] I added a `changelog.md` entry and rebuilt the package sources
according to [this short
guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

andrew-levy committed:
[expo-ui][ios] Allow nested ExpoSwiftUIViews (#34954)

Kudo committed:
[cli] fix dom components rendering error when running on tunnel mode (#34916)

# Why

having mixed content error when running dom components with tunnel

# How

serve bundle as schemeless `//host/bundle` url

EvanBacon committed:
Fix async bundling in development (#34986)

# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub
issues, forums posts, or feature requests.
-->

# How

<!--
How did you build this feature or fix this bug and why?
-->

# Test Plan

<!--
Please describe how you tested this change and how a reviewer could
reproduce your test, especially if this PR does not include automated
tests! If possible, please also provide terminal output and/or
screenshots demonstrating your test/reproduction.
-->

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [ ] I added a `changelog.md` entry and rebuilt the package sources
according to [this short
guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

---------

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

EvanBacon committed:
feat(metro-config): Add support for `require.resolveWeak` with tree shaking. (#34988)

# Why

- `require.resolveWeak` was being stripped during optimize pass and
therefore the results were not included in the final production bundle
when production tree shaking was enabled.

---------

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

gabrieldonadel committed:
[modules-core] Fix UnimplementedExpoView on macOS (#35014)

# Why

Building modules-core for macOS fails because UILabel is not available 

# How

Creates a wrapper class on top of NSTextField that mimics the basic
functionality of UILabel

# Test Plan

Run BareExpo locally 

<img width="1392" alt="image"
src="https://github.com/user-attachments/assets/6610d209-fe89-4323-b752-f92895e0f412"
/>


# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [ ] I added a `changelog.md` entry and rebuilt the package sources
according to [this short
guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
